code/notificationbuilder/src/main/java/com/adobe/marketing/mobile/notificationbuilder/internal/builders/TimerNotificationBuilder.kt

```diff
@@ -0,0 +1,188 @@
+
+        val notificationManager =
+            context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
+
+        // create the notification channel if needed
+        val channelIdToUse = notificationManager.createNotificationChannelIfRequired(context, template)
+
+        // check if the template is expired
+        val isExpired = template.isExpired()
+        val (smallLayout, expandedLayout) = initializeLayouts(context, isExpired)
+
+        // create the notification builder with the common settings applied
+        val notificationBuilder = AEPPushNotificationBuilder.construct(
+            context,
+            template,
+            channelIdToUse,
+            trackerActivityClass,
+            smallLayout,
+            expandedLayout,
+            R.id.basic_expanded_layout
+        )
+
+        // create the timer content according to the expiry time
+        val timerContent = if (isExpired) {
+            TimerContent(template.alternateTitle, template.alternateBody, template.alternateExpandedBody, template.alternateImage)
+        } else {
+            TimerContent(template.title, template.body, template.expandedBodyText, template.imageUrl)
+        }
+
+        // add text to collapsed layout
+        smallLayout.setTextViewText(R.id.notification_title, timerContent.title)
+        smallLayout.setTextViewText(R.id.notification_body, timerContent.body)
+
+        // add text to expanded layout
+        expandedLayout.setTextViewText(R.id.notification_title, timerContent.title)
+        expandedLayout.setTextViewText(R.id.notification_body_expanded, timerContent.expandedBody)
+
+        // download the image for expanded layout
+        val downloadedImageCount = PushTemplateImageUtils.cacheImages(listOf(timerContent.imageUrl))
+        if (downloadedImageCount == 0) {
+            Log.trace(LOG_TAG, TAG, "Timer push template unable to download the image for url : " + timerContent.imageUrl)
+            expandedLayout.setViewVisibility(R.id.expanded_template_image, View.GONE)
+        } else {
+            expandedLayout.setImageViewBitmap(R.id.expanded_template_image, PushTemplateImageUtils.getCachedImage(timerContent.imageUrl))
+        }
+
+        if (!isExpired) {
+            // set the timer clock
+            setTimerClock(smallLayout, expandedLayout, template)
+
+            // create the intent for the timer expiry
+            val intent = createIntent(template)
+            broadcastReceiverClass?.let {
+                intent.setClass(context, broadcastReceiverClass)
+            }
+
+            // create the pending intent for the timer expiry
+            val pendingIntent = PendingIntent.getBroadcast(context, 0, intent, PendingIntent.FLAG_MUTABLE or PendingIntent.FLAG_UPDATE_CURRENT)
+
+            // set the alarm manager to trigger the intent at the expiry time
+            val alarmManager = context.getSystemService(Context.ALARM_SERVICE) as AlarmManager
```

Comment:
we should verify exact alarms are allowed before setting it. for remind later notifications i used this code:
```kotlin
val alarmManager = context.getSystemService(Context.ALARM_SERVICE) as? AlarmManager
if (alarmManager != null) {
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S && alarmManager.canScheduleExactAlarms()) {
        Log.trace(LOG_TAG, SELF_TAG, "Exact alarms are permitted, scheduling an exact alarm for the current notification.")
        alarmManager.setExactAndAllowWhileIdle(AlarmManager.RTC_WAKEUP, triggerTime, pendingIntent)
    } else {
        // schedule an inexact alarm for the current notification
        Log.trace(LOG_TAG, SELF_TAG, "Exact alarms are not permitted, scheduling an inexact alarm for the current notification.")
        alarmManager.set(AlarmManager.RTC_WAKEUP, calendar.timeInMillis, pendingIntent)
    }
    // cancel the displayed notification
    notificationManager.cancel(tag.hashCode())
}
```

Reply:
As discussed offline, since we have timer counting down here and inexact alarm delivers within one hour window, it would not be accurate. We're going to try using foreground service for this case. 
cc: @PravinPK @prudrabhat

Reply:
Final decision: Only show this notification is exact permission is granted by the user. We need to document this additional step of app developer needing to request permission in the docs